### DISCUSSION TITLE
Weird bug with default compiler selection?

### DISCUSSION THREAD

**Discussion by kubatdkt** [OpenModelica/OpenModelica#10159] [Q&A] [answered]
Hello, 

I have a weird bug with minGW and compiler selection. In OMedit, if I don't choose gcc from the dropdown in Options>Simulate, the default option is "FR" and if I try to run a simulation it fails with:

> ....
> mingw32-make: *** [<builtin>: BouncingBall.o] Error 127
> mingw32-make: *** Waiting for unfinished jobs....
> /usr/bin/sh: FR: command not found
> mingw32-make: *** [<builtin>: BouncingBall_functions.o] Error 127
> /usr/bin/sh: FR: command not found
> mingw32-make: *** [<builtin>: BouncingBall_records.o] Error 127
> /usr/bin/sh: FR: command not found
> mingw32-make: *** [<builtin>: BouncingBall_01exo.o] Error 127
> mingw32-make: Leaving directory "H:/test"

As you see the issue is weird, because I'm on Windows 10 and don't have "/usr/bin/sh"

In OMedit, it's not a problem as I can just select gcc in the options drop down menu.

However, for me it becomes an issues as I'm trying to use **OMpython** and when I run a sample simulation, the same error appears:

> from OMPython import OMCSessionZMQ
> omc = OMCSessionZMQ()
> model_path=omc.sendExpression("getInstallationDirectoryPath()") + "/share/doc/omc/testmodels/"
> from OMPython import ModelicaSystem
> mod=ModelicaSystem(model_path + "BouncingBall.mo","BouncingBall", commandLineOptions="--target=gcc")

> Error: Error building simulator. Build log: mingw32-make: Entering directory 'H:/test
> FR  -Os -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse     -I"C:/OpenModelica/include/omc/c" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=BouncingBall -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=0 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0  -c -o BouncingBall.o BouncingBall.c
> FR  -Os -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse     -I"C:/OpenModelica/include/omc/c" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=BouncingBall -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=0 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0  -c -o BouncingBall_functions.o BouncingBall_functions.c
> FR  -Os -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse     -I"C:/OpenModelica/include/omc/c" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=BouncingBall -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=0 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0  -c -o BouncingBall_records.o BouncingBall_records.c
> /usr/bin/sh: FR: command not found
> FR  -Os -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse     -I"C:/OpenModelica/include/omc/c" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=BouncingBall -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=0 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0  -c -o BouncingBall_01exo.o BouncingBall_01exo.c
> mingw32-make: *** [<builtin>: BouncingBall.o] Error 127
> mingw32-make: *** Waiting for unfinished jobs....
> /usr/bin/sh: FR: command not found
> mingw32-make: *** [<builtin>: BouncingBall_functions.o] Error 127
> /usr/bin/sh: FR: command not found
> mingw32-make: *** [<builtin>: BouncingBall_records.o] Error 127
> /usr/bin/sh: FR: command not found
> mingw32-make: *** [<builtin>: BouncingBall_01exo.o] Error 127
> mingw32-make: Leaving directory 'H:/test'
> RESULT: 2

Here's the makefile. You will note that "CC=FR"

> \# Makefile generated by OpenModelica
> \# Platform: win64
> 
> \# Simulations use -O3 by default
> CC=FR
> CXX=clang++
> LINK=clang++ -shared -Xlinker --export-all-symbols
> EXEEXT=.exe
> DLLEXT=.dll
> CFLAGS_BASED_ON_INIT_FILE=
> \# define OMC_CFLAGS_OPTIMIZATION env variable to your desired optimization level to override this
> OMC_CFLAGS_OPTIMIZATION=-Os
> DEBUG_FLAGS=$(OMC_CFLAGS_OPTIMIZATION)
> CFLAGS=$(CFLAGS_BASED_ON_INIT_FILE) $(DEBUG_FLAGS) -Wno-parentheses-equality -falign-functions -mstackrealign -msse2 -mfpmath=sse ${MODELICAUSERCFLAGS}   
> CPPFLAGS= -I"C:/OpenModelica/include/omc/c" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=BouncingBall -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=0 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0
> \# define OMC_LDFLAGS_LINK_TYPE env variable to "static" to override this
> OMC_LDFLAGS_LINK_TYPE=dynamic
> RUNTIME_LIBS= -Wl,-B$(OMC_LDFLAGS_LINK_TYPE) -lSimulationRuntimeC -Wl,-Bdynamic  -Wl,-B$(OMC_LDFLAGS_LINK_TYPE) -lomcgc -lregex -ltre -lintl -liconv -lexpat -static-libgcc -luuid -loleaut32 -lole32 -limagehlp -lws2_32 -llis -lsundials_nvecserial -lsundials_sunmatrixdense -lsundials_sunmatrixsparse -lsundials_sunlinsoldense -lsundials_sunlinsolklu -lsundials_sunlinsollapackdense -lsundials_sunlinsolspbcgs -lsundials_sunlinsolspfgmr -lsundials_sunlinsolspgmr -lsundials_sunlinsolsptfqmr -lsundials_sunnonlinsolnewton -lsundials_cvode -lsundials_cvodes -lsundials_idas -lsundials_kinsol -lumfpack -lklu -lcolamd -lbtf -lamd -lsuitesparseconfig -lipopt -lcoinmumps -lpthread -lm  -lgfortran -lquadmath -lmingw32 -lgcc_eh -lmoldname -lmingwex  -luser32 -lkernel32 -ladvapi32 -lshell32 -lopenblas -lcminpack -Wl,-Bdynamic -lwsock32  -Wl,-B$(OMC_LDFLAGS_LINK_TYPE)  -lstdc++ -Wl,-Bdynamic 
> LDFLAGS=-L"C:/OpenModelica/lib//omc" -L"C://OpenModelica/lib" -Wl,--stack,16777216,-rpath,"C:/OpenModelica/lib//omc" -Wl,-rpath,"C:/OpenModelica/lib"  -fopenmp -Wl,-Bstatic -lregex -ltre -lintl -liconv -lexpat -lomcgc -lpthread -loleaut32 -limagehlp -lhdf5 -lz -lszip -Wl,-Bdynamic $(RUNTIME_LIBS) 
> DIREXTRA=-L"C:/OpenModelica/share/doc/omc/testmodels"
> MAINFILE=BouncingBall.c
> MAINOBJ=BouncingBall.o
> CFILES=BouncingBall_functions.c BouncingBall_records.c \
> BouncingBall_01exo.c BouncingBall_02nls.c BouncingBall_03lsy.c BouncingBall_04set.c BouncingBall_05evt.c BouncingBall_06inz.c BouncingBall_07dly.c \
> BouncingBall_08bnd.c BouncingBall_09alg.c BouncingBall_10asr.c BouncingBall_11mix.c BouncingBall_12jac.c BouncingBall_13opt.c BouncingBall_14lnz.c \
> BouncingBall_15syn.c BouncingBall_16dae.c BouncingBall_17inl.c BouncingBall_18spd.c 
> 
> OFILES=$(CFILES:.c=.o)
> GENERATEDFILES=$(MAINFILE) BouncingBall.makefile BouncingBall_literals.h BouncingBall_functions.h $(CFILES)
> 
> .PHONY: omc_main_target clean bundle
> 
> \# This is to make sure that BouncingBall_*.c are always compiled.
> .PHONY: $(CFILES)
> 
> omc_main_target: $(MAINOBJ) BouncingBall_functions.h BouncingBall_literals.h $(OFILES)
> 	$(CC) -I. -o BouncingBall$(EXEEXT) $(MAINOBJ) $(OFILES) $(CPPFLAGS) $(DIREXTRA)   $(CFLAGS) $(CPPFLAGS) $(LDFLAGS)
> clean:
> 	@rm -f BouncingBall_records.o $(MAINOBJ)
> 
> bundle:
> 	@tar -cvf BouncingBall_Files.tar $(GENERATEDFILES)

I've searched all around, but can't seem to find the option to force minGW to use GCC instead of this strange "FR" option. I've tried different versions of OM, but the error still persists.

Does anyone have an idea what to do? 
Thanks in advance!

**Comment by casella** [comment]
Maybe @mahge can help here?

**Reply by AnHeuermann** [reply]
> As you see the issue is weird, because I'm on Windows 10 and don't have "/usr/bin/sh"

OpenModelica ships its own Linux-like environment MINGW (Usually inside `C:\Program Files\OpenModelica1.21.0-dev-64bit\tools\msys`)  which has a lot of Linux tools like `sh`, `gcc`, `clang` we use for compiling C code on Windows.

![grafik](https://user-images.githubusercontent.com/38031952/217200591-e9471e7d-70b8-4358-9a46-8904d7c0a5f6.png)

You can change the tool-chain (`Target Build`) from default `MINGW` to one of the Microsoft ones, e.g. msvc19. The Compiler options will change depending on the tool-chain. For MINGW we support gcc and clang (and the C++ versions of that tools). In your case `clang` would be the correct one matching `clang++`.

I'm not sure where `FR` comes from. OMEdit and other tools read it from some configuration file, but I don't know where it is on Windows.
What version of OpenModelica do you have? (OMEdit Help->About OMedit) and did you try updating OpenModelica / re-installing OpenModelica?

**Reply by kubatdkt** [reply]
Hi, thanks for your reply. In fact it was an issue related to a random environment variable in my windows installation (corporate IT policy). See my reply to @adrpo below!

**Comment by adrpo** [comment, marked as answer]
You seems to have an environment variable called CC, I think that's why you have `CC=FR` in the Makefile.

**Reply by kubatdkt** [reply]
Thanks so much for your reply. In fact, there was an environmental variable CC=fr due to the corporate IT policy. Changing it to "GCC" did nothing, but after reinstallation of OpenModelica, GCC was chosen as the default both in OMedit and in OMPython !

It seems strange though, why upon installation OpenModelica looks for this environment variable. If MinGW is included by default, shouldn't it be "hardcoded" to have cc=gcc?

In any case, I will leave this thread for others if they experience this type of bug.

Thanks again!!
